libvirt/utils_volume.go

```diff
@@ -8,51 +8,46 @@ import (
 	"strings"
 	"time"
 
-	"github.com/libvirt/libvirt-go"
+	libvirt "github.com/digitalocean/go-libvirt"
 )
 
-func newCopier(virConn *libvirt.Connect, volume *libvirt.StorageVol, size uint64) func(src io.Reader) error {
+func newCopier(virConn *libvirt.Libvirt, volume *libvirt.StorageVol, size uint64) func(src io.Reader) error {
 	copier := func(src io.Reader) error {
-		var bytesCopied int64
+		r, w := io.Pipe()
+		defer w.Close()
 
-		stream, err := virConn.NewStream(0)
-		if err != nil {
-			return err
-		}
+		go func() error {
+			bytesCopied, err := io.Copy(w, src)
 
-		defer func() {
-			stream.Free()
-		}()
+			// if we get unexpected EOF this mean that connection was closed suddently from server side
+			// the problem is not on the plugin but on server hosting currupted images
+			if err == io.ErrUnexpectedEOF {
+				return fmt.Errorf("Error: transfer was unexpectedly closed from the server while downloading. Please try again later or check the server hosting sources")
+			}
+			if err != nil {
+				return fmt.Errorf("Error while copying source to volume %s", err)
+			}
 
-		if err := volume.Upload(stream, 0, size, 0); err != nil {
-			stream.Abort()
-			return fmt.Errorf("Error while uploading volume %s", err)
-		}
+			log.Printf("%d bytes uploaded\n", bytesCopied)
+			if uint64(bytesCopied) != size {
+				return fmt.Errorf("Error during volume Upload. BytesCopied: %d != %d volume.size", bytesCopied, size)
+			}
 
-		sio := NewStreamIO(*stream)
+			return w.Close()
 
-		bytesCopied, err = io.Copy(sio, src)
-		// if we get unexpected EOF this mean that connection was closed suddently from server side
-		// the problem is not on the plugin but on server hosting currupted images
-		if err == io.ErrUnexpectedEOF {
-			stream.Abort()
-			return fmt.Errorf("Error: transfer was unexpectedly closed from the server while downloading. Please try again later or check the server hosting sources")
-		}
-		if err != nil {
-			stream.Abort()
-			return fmt.Errorf("Error while copying source to volume %s", err)
-		}
+		}()
 
-		log.Printf("%d bytes uploaded\n", bytesCopied)
-		if uint64(bytesCopied) != size {
-			stream.Abort()
-			return fmt.Errorf("Error during volume Upload. BytesCopied: %d != %d volume.size", bytesCopied, size)
-		}
+		/*
+			* FIXME: use alternate simpler implementation without pipe?
```

Comment:
Because Digital Ocean library function takes an `io.Reader` to `StorageVolUpload(...)` I think we can avoid using pipe or any of the other commands previously required for streams.

Have validated this simpler implementation works for cloudinit test at least and will come back to as progress tests.

Going a step further, we can potentially rewrite `newCopier(...)` or drop and use `StorageVolUpload` inline. TBC.

Reply:
great

Reply:
Choose an implementation here, pipe or simple.